### ISSUE TITLE
freqs_cos data type of Llama3.2-1B mismatch with activation

### ISSUE THREAD

**Issue by WeiMa01** [pytorch/executorch#8614]
Why the llama in the llama model implemented by executorch is not converted to the xq_r type when rope is calculated. If the overridden model type = float16 is set when the llama3.2 - 1b guide pte is set, this will result in a mismatch between the node types of the rope's multiple nodes of the turned pte model. one xq_r is float32, one freqs_cos is float16, unable to run inference

    def apply_rotary_emb():
      xq_r, xq_i = xq.float().reshape(xq.shape[:-1] + (-1, 2)).unbind(-1)
      xk_r, xk_i = xk.float().reshape(xk.shape[:-1] + (-1, 2)).unbind(-1)
      freqs_cos = reshape_for_broadcast(freqs_cos, xq_r)
      freqs_sin = reshape_for_broadcast(freqs_sin, xq_r)
  
      xq_out_r = xq_r * freqs_cos - xq_i * freqs_sin
      xq_out_i = xq_r * freqs_sin + xq_i * freqs_cos
      xk_out_r = xk_r * freqs_cos - xk_i * freqs_sin
      xk_out_i = xk_r * freqs_sin + xk_i * freqs_cos
  
      xq_out = torch.stack([xq_out_r, xq_out_i], dim=-1).flatten(3)
      xk_out = torch.stack([xk_out_r, xk_out_i], dim=-1).flatten(3)
  
      return xq_out.type_as(xq), xk_out.type_as(xk)

**Comment by jackzhxng** [comment]
@WeiMa01 what is your export command / how are you running the model? `freqs_cos` should be fp16 when the dtype override `-d fp16`, not sure why `xq_r` is not fp16

**Comment by WeiMa01** [comment]
yes, we use -d fp16 to convert model to pte. thus freqs_cos should be fp16, which meet expection. however xq_r is float32, because it's activation which convert to fp32 when running rope, as I mentioned source code：
`xq_r, xq_i = xq.float().reshape(xq.shape[:-1] + (-1, 2)).unbind(-1)`
 This causes two input types not same during creating multiply node are created when inference pte
error log：
_@@@@@xnn_define_multiply2
@@@@@input1 dtype: FP32
@@@@@ input1_value->shape.num_dims: 4
@@@@@input1 shape: 1
@@@@@input1 shape: 127
@@@@@input1 shape: 32
@@@@@input1 shape: 32
@@@@@input2 dtype: FP16
@@@@@ input2_value->shape.num_dims: 4
@@@@@input2 shape: 1
@@@@@input2 shape: 127
@@@@@input2 shape: 1
@@@@@input2 shape: 32
@@@@@8, ERROR: xnn_subgraph_check_datatype_matches_two_inputs_

**Comment by jackzhxng** [comment]
Oh I see, I missed the `.float()`. This shouldn't error out for for portable lib, but it doesn't seem ideal to be doing dtype promotion for fp16? @kimishpatel 

For your situation specifically - seems also like you are using xnnpack, and fp16 is not fully supported at the moment.

**Comment by kimishpatel** [comment]
@WeiMa01 wha tis your export command?

**Comment by WeiMa01** [comment]
@kimishpatel ， export command is in the following:
python -m examples.models.llama.export_llama   --model "llama3_2"   --checkpoint "/Llama-3.2-1B/original/consolidated.00.pth"   --params "/Llama-3.2-1B/original/params.json"   -kv   --use_sdpa_with_kv_cache   -X   --xnnpack-extended-ops  -d fp16   --metadata '{"get_bos_id":128000, "get_eos_ids":[128009, 128001]}'   --generate_etrecord --max_seq_length 256   --output_name="llama3_2_xnnpack_fp16.pte"

**Comment by mcr229** [comment]
hi yea we should likely ad something to warn that ahead of time that mixed dtypes (fp32, fp16) aren't allowed within the XNNPACK Backend, would this work with portable ops? i.e. are mixed dtype inputs allowed within portable ops?

**Comment by WeiMa01** [comment]
> [@WeiMa01](https://github.com/WeiMa01) can you try without ` --xnnpack-extended-ops`? And for xnnpack lowering [@mcr229](https://github.com/mcr229) and [@digantdesai](https://github.com/digantdesai)

If we without ` --xnnpack-extended-ops`, some ops, like linear cann't delegated by XNNPACK, this is not our expection. We want original model not qmodel especially linear can delegated by XNNPACK, thus we add ` --xnnpack-extended-ops`

**Comment by WeiMa01** [comment]
> hi yea we should likely ad something to warn that ahead of time that mixed dtypes (fp32, fp16) aren't allowed within the XNNPACK Backend, would this work with portable ops? i.e. are mixed dtype inputs allowed within portable ops?

Hi, I also want to know when we use bf16 to export Llama3.2-1B, there also exists mixed dtypes(xq_r=fp32, freqs_cos=bf16), But this scene can create multiply node successful(two inputs dtypes are also fp32 ). where they convert freqs_cos from bf16 to fp32?

**Comment by mcr229** [comment]
bf16 support is still in progress. hard to estimate an exact timeline for this. The problem seems to arise from mixed dtypes of the mul node. You could try delegating everything but Mul as a workaround.

commenting out this line: 
https://github.com/pytorch/executorch/blob/0163902b867d202077a1ee9066689a4c253fa353/backends/xnnpack/partition/config/__init__.py#L91

in the mean time we will try to put up a pr to check mixed dtypes for binary ops.

**Comment by WeiMa01** [comment]
> bf16 support is still in progress. hard to estimate an exact timeline for this

Yes, bf16 not supported by XNNPACK currently. So I guess in order to complete the multiply node creation, one xq_r must be fp32, and the other freqs_cos = BF 16 must be transferred to fp32 at some point. I want to know exactly where this type conversion was done.

**Comment by jackzhxng** [comment]
@WeiMa01 it's converted here https://github.com/pytorch/executorch/blob/main/examples/models/llama/export_llama_lib.py#L982

**Comment by WeiMa01** [comment]
> [@WeiMa01](https://github.com/WeiMa01) it's converted here https://github.com/pytorch/executorch/blob/main/examples/models/llama/export_llama_lib.py#L982

Sorry， not here. I use -d bf16 to export llama, then this position will convert to freqs_cos to bf16 in AOT.
I mean in runtime the freqs_cos dtype from bf16 to convert fp32 when creating multiply node, But I don't know where it converted in runtime.
the log:
```
@@@@@xnn_define_multiply2
@@@@@input1 dtype: FP32
@@@@@ input1_value->shape.num_dims: 4
@@@@@input1 shape: 1
@@@@@input1 shape: 127
@@@@@input1 shape: 32
@@@@@input1 shape: 32
@@@@@input2 dtype: FP32
@@@@@ input2_value->shape.num_dims: 4
@@@@@input2 shape: 1
@@@@@input2 shape: 127
@@@@@input2 shape: 1
@@@@@input2 shape: 32
```

**Comment by mcr229** [comment]
@WeiMa01, the dtypes we are using when creating the multiply node are determined AoT. Meaning when we serialize the multiply instruction, we look at the incoming edges to multiply and serialize the dtype we saw for both these edges. AoT we should be checking that both inputs are both the same.

I'm not sure what you mean by converted in runtime,  the conversion we do for -d fp16 or -d fp32 essentially changes the AoT graph so that the dtype of the edges throughout the execution graph are all either fp16 or all fp32. I don't believe there should be any fp16 --> fp32 conversions within the XNNPACK graph.

**Comment by jackzhxng** [comment]
@mcr229 I think the issue is that for -d fp16 or -d bf16, mixed types get introduced to the graph from rope [here](https://github.com/pytorch/executorch/blob/main/examples/models/llama/rope.py#L84) where we cast to fp32 so some of the edges in the graph always going to be fp32

**Comment by WeiMa01** [comment]
> [@WeiMa01](https://github.com/WeiMa01), the dtypes we are using when creating the multiply node are determined AoT. Meaning when we serialize the multiply instruction, we look at the incoming edges to multiply and serialize the dtype we saw for both these edges. AoT we should be checking that both inputs are both the same.
> 
> I'm not sure what you mean by converted in runtime, the conversion we do for -d fp16 or -d fp32 essentially changes the AoT graph so that the dtype of the edges throughout the execution graph are all either fp16 or all fp32. I don't believe there should be any fp16 --> fp32 conversions within the XNNPACK graph.

But I have already checked the datyape of AOT. Just before convert to pte. we print the freqs_cos data type, it's bf16.
and in runtime, we checked the dtype when creating multiply node, two inputs are fp32. Thus I suppose freqs_cos must be converted to fp32 in runtime, and maybe in build XNNPACK graph of runtime init step

**Comment by WeiMa01** [comment]
> [@mcr229](https://github.com/mcr229) I think the issue is that for -d fp16 or -d bf16, mixed types get introduced to the graph from rope [here](https://github.com/pytorch/executorch/blob/main/examples/models/llama/rope.py#L84) where we cast to fp32 so some of the edges in the graph always going to be fp32

yes, it's main reason. I find HF llama model, they will convert freqs_cos to same dtype of xq_r. But executorch not did. Thus executorch may exist mixed dtype when using -d fp16 or bf16

**Comment by kimishpatel** [comment]
Funny that even tunehttps://github.com/pytorch/torchtune/blob/3d745a28bedb40dadf9761e265662da86da04662/tests/torchtune/models/llama2/scripts/compare_attention.py#L113 does that. @WeiMa01 do you have pointer to HF's impl. Either way I feel maybe it is worth trying to figure out if rope can tolerate bf16/fp16

**Comment by jackzhxng** [comment]
HF doesn't use complex in their impl, also seems like [`view_as_complex`](https://pytorch.org/docs/stable/generated/torch.view_as_complex.html) doesn't actually work with bf16 / fp16, hence the `float()` here. So yeah short term workaround is to leave this undelegated, long term we should rewrite this function in a different way without the cast / support mixed dtype
